src/Common/tests/Tests/System/IO/PathInternal.Windows.Tests.cs

```diff
@@ -141,5 +140,102 @@ public void AreRootsEqual(string first, string second, StringComparison comparis
         }
+
+        public static TheoryData<string, int> GetRootLength_Data => new TheoryData<string, int>
+        {
+            { @"C:\git\corefx", 3},
+            { @"C:\git\.\", 3},
+            { @"C:\git\..\", 3},
+            { @"C:\git\..\..\", 3},
+            { @"C:\..\", 3},
+            { @"C:\", 3},
+            { @"C:", 2},
+            { @"C:git\corefx", 2},
+            { @"C:git\.\", 2},
+            { @"C:git\..\", 2},
+            { @"C:..\", 2},
+        };
+
+        public static TheoryData<string, int> GetRootLengthRooted => new TheoryData<string, int>
+        {
+            { @"\tmp", 1},
+            { @"\.", 1},
+            { @"\..", 1},
+            { @"\tmp\..\..", 1},
+            { @"\\", 2},
+            { @"\\\", 3},
+        };
+
+        [Theory,
+            MemberData(nameof(GetRootLength_Data))]
+        public void GetRootLength(string path, int length)
+        {
+            Assert.Equal(length, PathInternal.GetRootLength(path));
+            Assert.Equal(length + PathInternal.ExtendedPathPrefix.Length, PathInternal.GetRootLength(@"\\?\" + path));
+            Assert.Equal(length + PathInternal.ExtendedPathPrefix.Length, PathInternal.GetRootLength(@"\\.\" + path));
+        }
+
+        [Theory,
+    MemberData(nameof(GetRootLengthRooted))]
+        public void GetRootLength_Rooted(string path, int length)
+        {
+            Assert.Equal(length, PathInternal.GetRootLength(path));
+        }
+
+        public static TheoryData<string, int> GetRootLength_UNCData => new TheoryData<string, int>
+        {
+            { @"Server\Share\git\corefx", 13},
+            { @"Server\Share\git\.\", 13},
+            { @"Server\Share\git\..\", 13},
+            { @"Server\Share\git\..\..\", 13},
+            { @"Server\Share\..\", 13},
+            { @"Server\Share\", 13},
+            { @"Server\Share", 12},
+            { @"\\Server\Share", 14},
+            //{ @"\\Server\\Share", 15},
+            { @"\\\Server\Share", 15},
+            { @"Server\Share\", 13},
+            { @"Server\Share\\git", 13},
+        };
+
+        [Theory,
+            MemberData(nameof(GetRootLength_UNCData))]
+        public void GetRootLengthUnc(string path, int length)
+        {
+            Assert.Equal(length + 2, PathInternal.GetRootLength(@"\\" + path));
+            Assert.Equal(length + PathInternal.UncExtendedPrefixLength, PathInternal.GetRootLength(@"\\?\UNC\" + path));
+            Assert.Equal(length + PathInternal.UncExtendedPrefixLength, PathInternal.GetRootLength(@"\\.\UNC\" + path));
+        }
+
+        public static TheoryData<string, int> GetRootLength_DeviceData => new TheoryData<string, int>
+        {
+            { @"..\", 3},
+            { @"..\..\", 3},
+            { @"..\.\", 3},
+            { @"...\", 4},
+            { @".\", 2},
+            { @".\..\", 2},
+            { @"..\", 3},
+            { @"\.\", 3},
```

Comment:
I don't know if this is what we want. Ideally we want NormalizeDirectorySeparators to eat all "extra" separators once past the root and end up with a valid path that doesn't need normalized. `\\?\\C:\foo.txt` is an invalid path and cannot be opened. I believe `\\.\\C:\foo.txt` would get turned into `\??\C:\foo.txt` and work, but I'm not positive.

So I think device paths should give you just the length of the prefix if you hit a directory separator right after. I think that is a better choice, but if that makes things too complicated I wouldn't sweat it.